Native auth MFA

SammyO opened the pull request:
Support for email OTP based MFA in native authentication flows.

- Added MFA results and states to the SDK interface
- Business logic & API integration
- Custom error handling for MFA required error during calls to /token with grant_type=refresh_token
- Private preview warnings in console and documentation
- Unit and E2E production tests

MSAL common PR: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/2489

SammyO committed:
MFA SDK interface (#2158)

New states:
- AwaitingMFAState: returned as part of `SignInResult.MFARequiredResult`
- MFARequiredState: returned for all other MFA operations, such as
`getAuthMethods()` and `sendChallenge()`

New results:
- `SignInResult.MFARequired(nextState: AwaitingMFAState)`: returned when
/token returns `mfa_required` error
- `MFARequiredResult.VerificationRequired(nextState: MFARequiredState,
val codeLength: Int, ...)`: returned when /challenge has successfully
challenged the user's default or selected auth method.
- `MFARequiredResult.SelectionRequired(nextState: MFARequiredState,
authMethods: List<AuthMethod>)`: returned when /introspect is called and
returns a list of auth methods

See `SignInTest.testSignInMFASimple()` for an example scenario.

**Note**: all SDK responses are currently hardcoded and are not hooked
up to underlying business logic. This PR should therefore not be merged
into `dev`

SammyO committed:
Merge branch 'dev' into sammy/mfa-epic

# Conflicts:
#	common
#	msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInEmailPasswordTest.kt

SammyO committed:
MFA business logic (#2160)

Requires
https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/2158
to be merged first.

MSAL common PR:
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/2474

SammyO committed:
Add changelog

SammyO committed:
Merge remote-tracking branch 'origin/sammy/mfa-epic' into sammy/mfa-epic

fadidurah commented on msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java:
```diff
@@ -757,6 +767,192 @@ public static SignInSubmitPasswordCommandParameters createSignInSubmitPasswordCo
+     * @return Command parameter object
+     * @throws ClientException
+     */
+    public static MFASelectedChallengeCommandParameters createMFASelectedChallengeCommandParameters(
```
Might be good to have one common method `createMFASelectedChallengeCommandParameters` that takes in all the needed parameters, and just have other methods expect less parameters but pass null for other fields they don't pass. That way there's less code to maintain.

SammyO committed:
update submodule reference

Yuki-YuXin committed:
Native Auth Email OTP MFA bug fix  (#2173)

SammyO committed:
Add warnings to Javadoc + log warnings when MFA methods are used. (#2168)

SammyO committed:
Update SignInResult.kt

SammyO committed:
MFA E2E tests

SammyO committed:
Merge branch 'sammy/mfa-epic' into sammy/mfa-e2e-tests

SammyO committed:
Merge branch 'dev' into sammy/mfa-epic

# Conflicts:
#	common

SammyO committed:
E2E tests using JSON configuration. Update test config types with 2 new types for MFA

SammyO committed:
Update assertions

SammyO committed:
MFA E2E tests (#2176)

SammyO committed:
Update tests

SammyO committed:
Skip tests

SammyO committed:
Merge branch 'sammy/mfa-e2e-tests' into sammy/mfa-epic

SammyO committed:
Merge remote-tracking branch 'origin/sammy/mfa-epic' into sammy/mfa-epic

# Conflicts:
#	msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt

SammyO committed:
Remove import

SammyO committed:
Merge branch 'dev' into sammy/mfa-epic

SammyO committed:
update submodule reference

SammyO commented on msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java:
```diff
@@ -524,7 +534,7 @@ public static SignUpSubmitPasswordCommandParameters createSignUpSubmitPasswordCo
      * @param username email address of the user
      * @param password password of the user
-     * @param scopes
+     * @param scopes scopes requested during sign in flow
```
Add test for this

SammyO commented on msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java:
```diff
@@ -757,6 +767,192 @@ public static SignInSubmitPasswordCommandParameters createSignInSubmitPasswordCo
+     * @return Command parameter object
+     * @throws ClientException
+     */
+    public static MFAChallengeCommandParameters createMFADefaultChallengeCommandParameters(
```
Check return type: why the generic `MFAChallengeCommandParameters`?

SammyO commented on msal/src/main/java/com/microsoft/identity/nativeauth/AuthMethod.kt:
```diff
@@ -0,0 +1,99 @@
+    // Auth method login hint (e.g. user@contoso.com)
+    val loginHint: String,
+
+    // Auth method login hint (email, etc.)
```
Comment is wrong

SammyO commented on msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/results/MFAResult.kt:
```diff
@@ -0,0 +1,73 @@
+
+/**
+ * Results related to MFA submit challenge operation, produced by
+ *  * [com.microsoft.identity.nativeauth.statemachine.states.MFARequiredState.submitChallenge]
```
typo

SammyO commented on msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/states/MFAStates.kt:
```diff
@@ -0,0 +1,627 @@
+                    }
+                }
+            } catch (e: Exception) {
+                MFAError(
```
Don't use generic MFAError?

Yuki-YuXin replied:
Take 'is INativeAuthCommandResult.Redirect -> { SubmitCodeError' as an example. Looks like we don't have separate error for BrowserRequired. Maybe MFAError?

SammyO replied:
@Yuki-YuXin what I meant is: https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/2185

SammyO committed:
Add scope check to tests

SammyO committed:
Merge branch 'dev' into sammy/mfa-epic

SammyO committed:
PR feedback

SammyO committed:
Introduce specific MFA errors (#2185)

- specific error classes makes sense when you have errors that can only
happen in one method. E.g. `invalidAttributesError()` should only be
available after `submitAttributes()`, i.e. we wrap that utility method
in `SignUpSubmitAttributesError` which is emitted from
`submitAttributes()`
- this has worked well so far: every method we have has method-specific
errors, so there has always been a reason to create dedicated error
classes
- however, for MFA this is not the case. Only `submitChallenge()` has a
specific error (invalid challenge), all the rest doesn't.
- however, we should still introduce specific MFA error classes anyway,
because without it it would either be a breaking change when we add it
(e.g. `requestChallenge()` will not longer emit `MFAError`, but now emit
`MFASubmitChallengeError`), or developers are going to need to check for
both error classes (e.g. `if error is MFAError || error is
MFASubmitChallengeError`)

SammyO committed:
Merge branch 'dev' into sammy/mfa-epic

SammyO committed:
update submodule reference

SammyO committed:
update submodule reference

nilo-ms commented on msal/src/main/java/com/microsoft/identity/nativeauth/AuthMethod.kt:
```diff
@@ -0,0 +1,99 @@
+    val id: String,
+
+    // Auth method challenge type (oob, etc.)
+    val challengeType: String,
```
nit: FYI on iOS we create a class with some util method like "isChallengeTyOOB() -> Bool".

SammyO replied:
We don't have that in any of our SDK states. It's a good idea, but I'd like to add that as a separate feature/improvement (also into existing flows). 
@Yuki-YuXin  FYI

nilo-ms commented on msal/src/main/java/com/microsoft/identity/nativeauth/AuthMethod.kt:
```diff
@@ -0,0 +1,99 @@
+    override fun toUnsanitizedString(): String = "AuthMethod(id=$id, " +
+            "challengeType=$challengeType, loginHint=$loginHint, challengeChannel=$challengeChannel)"
+
+    override fun toString(): String = "AuthMethod(id=$id, challengeChannel=$challengeChannel)"
```
nit: is there a specific reason toString doesn't print "loginHint" and "challengeType"?

SammyO replied:
loginHint is PII. Challenge type probably not.

SammyO replied:
Updated

nilo-ms commented on msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/states/MFAStates.kt:
```diff
@@ -0,0 +1,627 @@
+     * <strong><u>Warning: this API is experimental. It may be changed in the future without notice. Do not use in production applications.</u></strong>
+     * @return The result of the submit challenge action.
+     */
+    suspend fun submitChallenge(challenge: String): MFASubmitChallengeResult {
```
on iOS we check that the challenge is not empty. If it is, we return "invalidChallenge" error

SammyO replied:
We don't do that kind of validation in the SDK layer (also not in submitCode in sign in, sign up and SSPR flows). We do this in the API layer (`SignInTokenRequest`), where we're re-using the existing signInSubmitCode flows.

nilo-ms commented on msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/states/MFAStates.kt:
```diff
@@ -0,0 +1,627 @@
+                    }
+                    is SignInCommandResult.IncorrectCode -> {
+                        MFASubmitChallengeError(
+                            errorType = ErrorTypes.INVALID_CODE,
```
Is the error type name "invalid code" or "invalid challenge"? On iOS we used "invalid challenge" because the user submits a challenge and not a code

SammyO replied:
Updated

SammyO committed:
PR feedback

SammyO committed:
Merge branch 'dev' into sammy/mfa-epic

# Conflicts:
#	common

SammyO committed:
update submodule

SammyO committed:
PR feedback
